File: code/agents.py
"""
Vectorised agent pool.  All agent attributes are numpy arrays so every
per-agent operation can be expressed as a single array expression.
"""

from __future__ import annotations

import numpy as np
from enum import IntEnum

from config import CONFIG


#  Health states 

class HealthState(IntEnum):
    SUSCEPTIBLE             = 0
    EXPOSED                 = 1
    INFECTIOUS_ASYMPTOMATIC = 2
    INFECTIOUS_SYMPTOMATIC  = 3
    RECOVERED               = 4
    DEAD                    = 5


#  AgentPool 

class AgentPool:
    """
    Stores all per-agent state as numpy arrays of length n.

    Having every attribute as a flat array lets us apply boolean masks and
    arithmetic without any Python-level loops over agents.
    """

    def __init__(self, n: int, rng: np.random.Generator) -> None:
        self.n   = n
        self.rng = rng

        #  Health 
        self.health_state  = np.full(n, HealthState.SUSCEPTIBLE, dtype=np.int8)
        self.days_in_state = np.zeros(n, dtype=np.int16)
        self.is_alive      = np.ones(n,  dtype=bool)
        self.vaccinated    = np.zeros(n, dtype=bool)

        #  Economics 
        self.wallet = rng.normal(
            CONFIG["INITIAL_WALLET_MEAN"],
            CONFIG["INITIAL_WALLET_STD"],
            n,
        ).clip(min=CONFIG["INITIAL_WALLET_MIN"])

        self.base_consumption = rng.normal(
            CONFIG["BASE_CONSUMPTION_MEAN"],
            CONFIG["BASE_CONSUMPTION_STD"],
            n,
        ).clip(min=CONFIG["BASE_CONSUMPTION_MIN"])

        # 0 = never breaks lockdown; 1 = always breaks lockdown
        self.risk_tolerance = rng.uniform(0.0, 1.0, n)

        #  Employment 
        # -1 = unemployed
        self.company_id = np.full(n, -1, dtype=np.int32)
        self.employed   = np.zeros(n, dtype=bool)

    #  Boolean masks 

    def susceptible_mask(self) -> np.ndarray:
        return (
            (self.health_state == HealthState.SUSCEPTIBLE)
            & self.is_alive
            & ~self.vaccinated
        )

    def infectious_mask(self) -> np.ndarray:
        return (
            (
                (self.health_state == HealthState.INFECTIOUS_ASYMPTOMATIC)
                | (self.health_state == HealthState.INFECTIOUS_SYMPTOMATIC)
            )
            & self.is_alive
        )

    def at_work_mask(self, lockdown_level: int) -> np.ndarray:
        """
        Returns True for every agent who physically goes to work this tick.

        Symptomatic agents mostly self-isolate but a small fraction still
        attend (poverty / income pressure). Under lockdown the poverty trap
        overrides compliance; otherwise agents comply proportionally to the
        inverse of their risk_tolerance scaled by lockdown stringency.
        """
        symp = self.health_state == HealthState.INFECTIOUS_SYMPTOMATIC
        non_symp = (
            self.is_alive
            & self.employed
            & ~symp
            & (self.health_state != HealthState.DEAD)
        )
        symp_attends = (
            self.is_alive
            & self.employed
            & symp
            & (self.rng.random(self.n) < CONFIG["SYMPTOMATIC_WORK_ATTENDANCE_FACTOR"])
        )
        can_work = non_symp | symp_attends

        if lockdown_level == 0:
            return can_work

        poverty_trap = self.wallet < (
            self.base_consumption * CONFIG["POVERTY_TRAP_MULTIPLIER"]
        )
        # Higher lockdown level → smaller effective break probability
        break_prob = self.risk_tolerance / lockdown_level
        breaks_by_choice = self.rng.random(self.n) < break_prob
        return can_work & (poverty_trap | breaks_by_choice)

    def at_market_mask(self, lockdown_level: int) -> np.ndarray:
        """
        Returns True for agents who visit markets / public spaces this tick.
        Symptomatic agents mostly self-isolate but a small fraction still
        venture out for essentials.
        """
        symp = self.health_state == HealthState.INFECTIOUS_SYMPTOMATIC
        non_symp = (
            self.is_alive
            & ~symp
            & (self.health_state != HealthState.DEAD)
        )
        symp_attends = (
            self.is_alive
            & symp
            & (self.rng.random(self.n) < CONFIG["SYMPTOMATIC_MARKET_ATTENDANCE_FACTOR"])
        )
        mobile = non_symp | symp_attends

        if lockdown_level == 0:
            return mobile

        poverty_trap    = self.wallet < (
            self.base_consumption * CONFIG["POVERTY_TRAP_MULTIPLIER"]
        )
        compliance_rate = lockdown_level * 0.20
        goes_out        = poverty_trap | (self.rng.random(self.n) > compliance_rate)
        return mobile & goes_out

    #  Aggregate statistics 

    @property
    def total_alive(self) -> int:
        return int(self.is_alive.sum())

    @property
    def total_dead(self) -> int:
        return int((~self.is_alive).sum())

    @property
    def unemployment_rate(self) -> float:
        alive = int(self.is_alive.sum())
        if alive == 0:
            return 1.0
        return float((self.is_alive & ~self.employed).sum() / alive)

    @property
    def mean_wallet(self) -> float:
        alive = self.is_alive
        return float(self.wallet[alive].mean()) if alive.any() else 0.0

    @property
    def median_wallet(self) -> float:
        alive = self.is_alive
        return float(np.median(self.wallet[alive])) if alive.any() else 0.0

    def state_counts(self) -> dict[str, int]:
        hs = self.health_state
        return {
            "susceptible":              int((hs == HealthState.SUSCEPTIBLE).sum()),
            "exposed":                  int((hs == HealthState.EXPOSED).sum()),
            "infectious_asymptomatic":  int((hs == HealthState.INFECTIOUS_ASYMPTOMATIC).sum()),
            "infectious_symptomatic":   int((hs == HealthState.INFECTIOUS_SYMPTOMATIC).sum()),
            "recovered":                int((hs == HealthState.RECOVERED).sum()),
            "dead":                     int((hs == HealthState.DEAD).sum()),
        }

File: web/src/sim/agents.ts
// Vectorised agent pool — every per-agent attribute is a typed array so we
// can iterate the population in tight loops without GC pressure.

import { CONFIG } from './config'
import { PRNG } from './prng'

export const HealthState = {
  Susceptible: 0,
  Exposed: 1,
  InfectiousAsymptomatic: 2,
  InfectiousSymptomatic: 3,
  Recovered: 4,
  Dead: 5,
} as const
export type HealthStateValue = (typeof HealthState)[keyof typeof HealthState]

export class AgentPool {
  readonly n: number
  readonly rng: PRNG

  // Health
  healthState: Int8Array
  daysInState: Int16Array
  /** 1 = alive, 0 = dead. Stored as Uint8Array for speed. */
  isAlive: Uint8Array
  vaccinated: Uint8Array

  // Economics
  wallet: Float32Array
  baseConsumption: Float32Array
  /** [0, 1] — agents above this are more likely to break lockdown. */
  riskTolerance: Float32Array

  // Employment
  /** -1 = unemployed. */
  companyId: Int32Array
  employed: Uint8Array

  constructor(n: number, rng: PRNG) {
    this.n = n
    this.rng = rng

    this.healthState = new Int8Array(n) // 0 = Susceptible by default
    this.daysInState = new Int16Array(n)
    this.isAlive = new Uint8Array(n).fill(1)
    this.vaccinated = new Uint8Array(n)

    this.wallet = rng.normal(
      CONFIG.INITIAL_WALLET_MEAN,
      CONFIG.INITIAL_WALLET_STD,
      n,
    )
    for (let i = 0; i < n; i++) {
      if (this.wallet[i] < CONFIG.INITIAL_WALLET_MIN) {
        this.wallet[i] = CONFIG.INITIAL_WALLET_MIN
      }
    }

    this.baseConsumption = rng.normal(
      CONFIG.BASE_CONSUMPTION_MEAN,
      CONFIG.BASE_CONSUMPTION_STD,
      n,
    )
    for (let i = 0; i < n; i++) {
      if (this.baseConsumption[i] < CONFIG.BASE_CONSUMPTION_MIN) {
        this.baseConsumption[i] = CONFIG.BASE_CONSUMPTION_MIN
      }
    }

    this.riskTolerance = rng.uniform(0, 1, n)

    this.companyId = new Int32Array(n).fill(-1)
    this.employed = new Uint8Array(n)
  }

  /** Returns a fresh Uint8Array mask: agents who attend their workplace this tick. */
  atWorkMask(lockdownLevel: number): Uint8Array {
    const n = this.n
    const out = new Uint8Array(n)
    const symFactor = CONFIG.SYMPTOMATIC_WORK_ATTENDANCE_FACTOR
    const povertyMul = CONFIG.POVERTY_TRAP_MULTIPLIER
    for (let i = 0; i < n; i++) {
      if (!this.isAlive[i] || !this.employed[i]) continue
      const sym = this.healthState[i] === HealthState.InfectiousSymptomatic
      let canWork = false
      if (sym) {
        if (this.rng.next() < symFactor) canWork = true
      } else {
        canWork = true
      }
      if (!canWork) continue
      if (lockdownLevel === 0) {
        out[i] = 1
        continue
      }
      const poverty = this.wallet[i] < this.baseConsumption[i] * povertyMul
      const breakProb = this.riskTolerance[i] / lockdownLevel
      const breaks = this.rng.next() < breakProb
      if (poverty || breaks) out[i] = 1
    }
    return out
  }

  /** Returns a fresh Uint8Array mask: agents who go to public spaces this tick. */
  atMarketMask(lockdownLevel: number): Uint8Array {
    const n = this.n
    const out = new Uint8Array(n)
    const symFactor = CONFIG.SYMPTOMATIC_MARKET_ATTENDANCE_FACTOR
    const povertyMul = CONFIG.POVERTY_TRAP_MULTIPLIER
    for (let i = 0; i < n; i++) {
      if (!this.isAlive[i]) continue
      const sym = this.healthState[i] === HealthState.InfectiousSymptomatic
      let mobile = false
      if (sym) {
        if (this.rng.next() < symFactor) mobile = true
      } else {
        mobile = true
      }
      if (!mobile) continue
      if (lockdownLevel === 0) {
        out[i] = 1
        continue
      }
      const poverty = this.wallet[i] < this.baseConsumption[i] * povertyMul
      const compliance = lockdownLevel * 0.20
      if (poverty || this.rng.next() > compliance) out[i] = 1
    }
    return out
  }

  // ---- aggregate stats ----

  totalAlive(): number {
    let count = 0
    for (let i = 0; i < this.n; i++) count += this.isAlive[i]
    return count
  }

  totalDead(): number {
    return this.n - this.totalAlive()
  }

  unemploymentRate(): number {
    let alive = 0
    let unemp = 0
    for (let i = 0; i < this.n; i++) {
      if (this.isAlive[i]) {
        alive++
        if (!this.employed[i]) unemp++
      }
    }
    return alive === 0 ? 1 : unemp / alive
  }

  meanWallet(): number {
    let sum = 0
    let count = 0
    for (let i = 0; i < this.n; i++) {
      if (this.isAlive[i]) {
        sum += this.wallet[i]
        count++
      }
    }
    return count === 0 ? 0 : sum / count
  }

  medianWallet(): number {
    const tmp: number[] = []
    for (let i = 0; i < this.n; i++) {
      if (this.isAlive[i]) tmp.push(this.wallet[i])
    }
    if (tmp.length === 0) return 0
    tmp.sort((a, b) => a - b)
    const m = tmp.length >> 1
    return tmp.length % 2 === 0 ? (tmp[m - 1] + tmp[m]) / 2 : tmp[m]
  }

  stateCounts(): {
    susceptible: number
    exposed: number
    infectiousAsymptomatic: number
    infectiousSymptomatic: number
    recovered: number
    dead: number
  } {
    const counts = [0, 0, 0, 0, 0, 0]
    for (let i = 0; i < this.n; i++) counts[this.healthState[i]]++
    return {
      susceptible: counts[HealthState.Susceptible],
      exposed: counts[HealthState.Exposed],
      infectiousAsymptomatic: counts[HealthState.InfectiousAsymptomatic],
      infectiousSymptomatic: counts[HealthState.InfectiousSymptomatic],
      recovered: counts[HealthState.Recovered],
      dead: counts[HealthState.Dead],
    }
  }
}
